### PR TITLE
Sign NOTIFY with TSIG

### LINKED ISSUE #656: Sign supermaster/superslave zone transfers with TSIG
Add support for signing superslave transfers. Currently, TSIG key is assigned to zone. Because superslave creates zones when notified, it doesn't have any key to check signatures. 

IP-based security is not strong enough, especially when using DNSSEC (superslave then becomes the weak link - attacker could do MITM and put his own zones in authoritative server).

### PR DESCRIPTION
**Pull request by cmouse**
Adds TSIG support for notify and also modifies supermaster to accept any valid TSIG key in notify and use that as TSIG axfr key for the domain. Closes #656.

TODO:
- [x] Documentation
- [x] Option to allow other products to work
- [x] Testing

### CODE REVIEW

**Comment by cmouse** [comment]
Ready for review, I have tested this myself. Ideas for how to test this are welcome.

**Comment by cmouse** [comment]
Few things:
1. This will *BREAK* pre-4.0 master, 4.0-slave interoperability if you use TSIG, as slaves insist after this that you use TSIG signed notifications.
2. supermaster provisioning will use *any* *valid* TSIG key for authorization and *map* that key to the new domain.

**Comment by cmouse** [comment]
Missing, documentation, needs pointers what to do document, and where
Missing, testing, but we don't test supermaster either.
Otherwise ready for review.

**Comment by cmouse** [comment]
Ready for review. Again.

**Comment by mind04** [comment]
This is not a "mine is bigger contest" please use minimal zones in your test.
TSIG is about not being able to transfer/create zones if the keys don't match. This aspect is completely missing in your test
And please don't squash,or rebase, commits after you have asked for a review. Our time is to valuable to do a compete review over and over again. There is plenty of time for squashing after reviewing is done.

**Comment by cmouse** [comment]
I'll minimize the zones. They were copied off the regression-tests so I forgot.

The aspect of "not being able to transfer" is tested elsewhere, also, TSIG is used to grant/deny access to tsig.com, which is tested and enforced by setting allow-axfr-ips to empty value. 

The rebase/squash was done to hide a small nit, not to waste your valuable time.

**Comment by mind04** [comment]
The pull title says "Sign NOTIFY with TSIG". Can you show me where this is tested?

File: pdns/packethandler.cc [hunk 1]
```diff
@@ -706,7 +706,7 @@ int PacketHandler::trySuperMaster(DNSPacket *p)
   }
 }
 
-int PacketHandler::trySuperMasterSynchronous(DNSPacket *p)
+int PacketHandler::trySuperMasterSynchronous(DNSPacket *p, const DNSName& tsigkeyname)
```
Comment (cmouse) [inline pdns/packethandler.cc]: it is tested here (part 1/2)

File: regression-tests.nobackend/supermaster/command [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+# then we check that we got records in database for both domains
+for domain in test.com tsig.com; do
+  reccount=`sqlite3 slave.db "select count(*) from records join domains on records.domain_id = domains.id where domains.name='$domain'"`
+  if [ $reccount -lt 3 ]; then
+    echo "Transfer of '$domain' did not succeed"
+  fi
+done
+
+# then check that tsig got provisioned
```
Comment (cmouse) [inline regression-tests.nobackend/supermaster/command]: and it is tested here (part 2/2)

**Comment by cmouse** [comment]
If you feel it is necessary, I can modify the test suite to ensure the next notification is also verified.

**Comment by mind04** [comment]
My point is you only test for success. The test does not ensure failing keys get rejected.
I like to see "Transfer of xxx did not succeed' lines in expected_results for zones without matching keys or no keys at all.

**Comment by cmouse** [comment]
True. But TSIG transfer success/failure is tested elsewhere, already present in the test suite. This test is about supermaster. 

I'll see if i can somehow test the notification failure (due to missing key) somehow in the main test suite where notify is tested as well.

I can add tsig transfer failure test to the main test suite in separate pr.

**Comment by mind04** [comment]
What happens on the slave if a initial notify for a new supermaster zone is signed with an unknown key (or no key if 'allow-insecure-notify=no')?

I hope the answer is "it get rejected, and no zone is created on the slave'

My question is, test this...

**Comment by cmouse** [comment]
A zone gets created because supermasters are controlled with supermasters table. Unknown/missing tsig key causes no association. But it has to come from valid supermaster.

**Comment by cmouse** [comment]
@mind04 it has now two tests, one for unsigned mode, one for signed mode. There are tests in place to ensure domains do not appear w/o TSIG etc. Can you look it thru once more now?

**Comment by mind04** [comment]
Wonderful an other rebase... I have no clue what you have changed since my last review. 
Sorry, i'm not going to spend more time on this pull doing full reviews over and over again.

**Comment by cmouse** [comment]
Sorry to hear that. Thank you for your effort! Hopefully someone else can review this.

File: pdns/dnspacket.cc [hunk 3]
```diff
@@ -667,3 +667,7 @@ bool checkForCorrectTSIG(const DNSPacket* q, UeberBackend* B, DNSName* keyname,
 
   return result;
 }
+
+const DNSName& DNSPacket::getTSIGKeyname() const {
```
Comment (ahupowerdns) [inline pdns/dnspacket.cc]: umf please no references to internals when we don't really need them

**Comment by cmouse** [comment]
Fixes #1044

**Comment by cmouse** [comment]
Updated to current master. Waiting for review.

File: docs/markdown/authoritative/modes-of-operation.md [hunk 4]
```diff
@@ -8,7 +8,7 @@ MySQL replication has proven to be very robust and well suited, even over transa
 To use native replication, configure your backend storage to do the replication and do not configure PDNS to do so.
 
 # Master operation
-When operating as a master, PDNS sends out notifications of changes to slaves, which react to these notifications by querying PDNS to see if the zone changed, and transferring its contents if it has. Notifications are a way to promptly propagate zone changes to slaves, as described in [RFC 1996](http://tools.ietf.org/html/rfc1996).
+When operating as a master, PDNS sends out notifications of changes to slaves, which react to these notifications by querying PDNS to see if the zone changed, and transferring its contents if it has. Notifications are a way to promptly propagate zone changes to slaves, as described in [RFC 1996](http://tools.ietf.org/html/rfc1996). Since version 4.0, the NOTIFY messages have a TSIG record added (transaction signature) if zone has been configured to use TSIG and feature has been enabled.
```
Comment (Habbie) [inline docs/markdown/authoritative/modes-of-operation.md]: 4.0.0, not 4.0

File: docs/markdown/authoritative/modes-of-operation.md [hunk 5]
```diff
@@ -43,6 +43,8 @@ Slave operation can also be programmed using several [`pdns_control`](internals.
 
 Since version 2.9.21, PowerDNS supports multiple masters. For the BIND backend, the native BIND configuration language suffices to specify multiple masters, for SQL based backends, list all master servers separated by commas in the 'master' field of the domains table.
 
+Since version 4.0, PowerDNS requires that slaves sign their notifications. During transition and interoperation with other nameservers, you can use options **allow-unsigned-notify** to permit unsigned notifications. For 4.0 this is turned off by default, but it might be turned on permanently in future releases.
```
Comment (Habbie) [inline docs/markdown/authoritative/modes-of-operation.md]: slaves do not send notifications; why would they sign them?

File: pdns/dnspacket.cc (hunk 3 above)
Comment (Habbie) [inline pdns/dnspacket.cc]: This was not fixed it appears?

File: regression-tests.nobackend/supermaster-signed/description [hunk 6]
```diff
@@ -0,0 +1 @@
+Tests that supermaster backend works and assings TSIG keyname to zone on provision
```
Comment (Habbie) [inline regression-tests.nobackend/supermaster-signed/description]: assigns

File: regression-tests.nobackend/supermaster-unsigned/description [hunk 7]
```diff
@@ -0,0 +1 @@
+Tests that supermaster backend works and assings TSIG keyname to zone on provision
```
Comment (Habbie) [inline regression-tests.nobackend/supermaster-unsigned/description]: assigns

**Comment by Habbie** [comment]
few nits; then LGTM

**Comment by cmouse** [comment]
Thanks. I'll look into these.

**Comment by cmouse** [comment]
Rebased over master, no squashing,

**Comment by cmouse** [comment]
Ready for merge or squash if wanted

**Pull request merged**